### PR TITLE
bpo-29854: Print readline version information when running test suite

### PR DESCRIPTION
**Pull request by berkerpeksag**

### CODE REVIEW

**Review by nirs — COMMENTED** [review]
:+1:

**Comment by vstinner** [comment]
I would prefer to dump the curses version in test_readline, as done by test_tcl or test_gdb in verbose mode. When a test fails on buildbots, the test is run again in verbose mode.

**Comment by vstinner** [comment]
I prefer to not make regrtest header too big.

**Comment by berkerpeksag** [comment]
Why would we want to print curses version in test_readline? Unless I'm missing something, curses doesn't have anything to do with bpo-29854.

tcl and gdb are much less common dependencies than readline. readline is probably available on all Unix platforms and I think it makes sense to easily see which version is installed on the host.

> I prefer to not make regrtest header too big.

Sorry, I don't have time for this kind of bikeshedding. We don't design a production level UI for end users and it's already hard to read outputs of regrrest and Buildbot. I prefer not to spend too much time to find which version of readline installed at http://buildbot.python.org/all/builders/x86%20Tiger%203.x/builds/908/steps/test/logs/stdio Plus, that header has already prints less useful information so I don't think two additional lines would set the world on fire :)

**Comment by vstinner** [comment]
> Why would we want to print curses version in test_readline? Unless I'm missing something, curses doesn't have anything to do with bpo-29854.

Oh sorry, I'm always confused between readline and curses, they are the same thing for me, sorry :-)

By the way, you should mention bpo-29854 in your PR title. It helps later to get the rationale of a change ;-)

> tcl and gdb are much less common dependencies than readline. readline is probably available on all Unix platforms and I think it makes sense to easily see which version is installed on the host.

Another issue with importing readline early is that importing a module has side effect, especially readline. See for example http://bugs.python.org/issue19884 and TestReadline.test_init(): this test still fails on our RHEL buildbot, so importing readline still have an annoying side effect of writing a string into stdout.

I would prefer to limit imports at regrtest startup, to make it more deterministic.


> I prefer not to spend too much time to find which version of readline installed at http://buildbot.python.org/all/builders/x86%20Tiger%203.x/builds/908/steps/test/logs/stdio

Would you mind to review my PR #2619? When it will be merged, I can retrieve the libreadline version for you and put it in the issue.

**Comment by vstinner** [comment]
We discuss this PR on IRC with @berkerpeksag. If imports are done are done in subprocess to prevent side effects, I'm ok to get such change merged ;-)

File: Lib/test/libregrtest/main.py [hunk 1]
```diff
@@ -432,6 +433,26 @@ def display_header(self):
         print("== encodings: locale=%s, FS=%s"
               % (locale.getpreferredencoding(False),
                  sys.getfilesystemencoding()))
+        rl_version_cmd = """if 1:
+        try:
+            import readline
+        except ImportError:
+            readline = None
+        if readline is not None:
```
Comment (vstinner) [inline Lib/test/libregrtest/main.py]: this line just can be an "else:" no?

File: Lib/test/libregrtest/main.py [hunk 2]
```diff
@@ -432,6 +433,26 @@ def display_header(self):
         print("== encodings: locale=%s, FS=%s"
               % (locale.getpreferredencoding(False),
                  sys.getfilesystemencoding()))
+        rl_version_cmd = """if 1:
+        try:
+            import readline
+        except ImportError:
```
Comment (vstinner) [inline Lib/test/libregrtest/main.py]: Maybe use sys.exit(1) to not try to decode empty stdout in the parent process.

File: Lib/test/libregrtest/main.py [hunk 3]
```diff
@@ -432,6 +433,26 @@ def display_header(self):
         print("== encodings: locale=%s, FS=%s"
               % (locale.getpreferredencoding(False),
                  sys.getfilesystemencoding()))
+        rl_version_cmd = """if 1:
```
Comment (vstinner) [inline Lib/test/libregrtest/main.py]: You may put the textwrap.dedent() directly here. But do you need "if 1:" in that case?

File: Lib/test/libregrtest/main.py [hunk 4]
```diff
@@ -432,6 +433,26 @@ def display_header(self):
...
+        rl_output = subprocess.run([sys.executable, "-c", textwrap.dedent(rl_version_cmd)],
+                                   stdout=subprocess.PIPE, stderr=subprocess.PIPE)
+        if rl_output.returncode == 0:
+            try:
+                rl_version, rl_implementation = rl_output.stdout.decode().splitlines()
+            except ValueError:
+                pass
+            else:
+                print(f"== readline version: {rl_version}")
+                print(f"== readline implementation: {rl_implementation}")
```
Comment (vstinner) [inline Lib/test/libregrtest/main.py]: can you please write both info on a single line?

File: Lib/test/libregrtest/main.py [hunk 5]
```diff
@@ -432,14 +433,26 @@ def display_header(self):
...
             import readline
         except ImportError:
             readline = None
         if readline is not None:
-            print(f"== readline version: {readline._READLINE_VERSION:#06x}")
+            print(f"{readline._READLINE_VERSION:#06x}")
             is_editline = readline.__doc__ and "libedit" in readline.__doc__
-            print("== readline implementation:", "editline" if is_editline else "GNU readline")
+            print("editline" if is_editline else "GNU readline")
+        """
```
Comment (nirs) [inline Lib/test/libregrtest/main.py]: It is very hard to read this code like this. The other program that you run should be indented differently from the other code.

Also the code to return the info should be in separate function, and here we can just print the values.

File: Lib/test/libregrtest/main.py (hunk 3 above)
Comment (berkerpeksag) [reply Lib/test/libregrtest/main.py]: Do you mean

```py
rl_version_cmd = textwrap.dedent("""
...
""")
```

?

And, you're right, we don't need `if 1:`, thanks!

File: Lib/test/libregrtest/main.py (hunk 2 above)
Comment (berkerpeksag) [reply Lib/test/libregrtest/main.py]: Good point!

File: Lib/test/libregrtest/main.py (hunk 4 above)
Comment (berkerpeksag) [reply Lib/test/libregrtest/main.py]: Sure, what do you thing about the following format?:

```
== readline: 0x0602 (GNU readline)
```

Comment (vstinner) [reply Lib/test/libregrtest/main.py]: I like it. Maybe just add version somewhere? "== readline: version 0x0602 (GNU readline)"

Comment (berkerpeksag) [reply Lib/test/libregrtest/main.py]: Done:

```
== readline: version 0x0602 (GNU readline)
```

Thanks!

File: Lib/test/libregrtest/main.py (hunk 5 above)
Comment (berkerpeksag) [reply Lib/test/libregrtest/main.py]: So what you're suggesting is similar to the following snippet, right?

```py
rl_version_cmd = """
def get_rl_info():
    # all code is here
    return version, implementation
version, implementation = get_rl_info()
print(version)
print(implementation)
"""
```

Comment (berkerpeksag) [reply Lib/test/libregrtest/main.py]: Or:

```py
rl_version_cmd = """
def get_rl_info():
    # all code is here
    return version, implementation
version, implementation = get_rl_info()
print("== readline: version {version} ({implementation})")
"""
```

**Comment by berkerpeksag** [comment]
I've addressed all review comments (and hopefully I got @nirs' comment right) Thanks!

File: Lib/test/libregrtest/main.py [hunk 6]
```diff
@@ -433,26 +433,23 @@ def display_header(self):
...
+        rl_version_cmd = textwrap.dedent("""
+            import sys
             try:
-                rl_version, rl_implementation = rl_output.stdout.decode().splitlines()
-            except ValueError:
-                pass
+                import readline
+            except ImportError:
+                sys.exit(1)
             else:
```
Comment (nirs) [inline Lib/test/libregrtest/main.py]: We can remove the try/else/exit, just import readline:

    import readline

If import fails, we exit with exit code 1 anyway, no special code needed.

File: Lib/test/libregrtest/main.py [hunk 7]
```diff
@@ -433,26 +433,23 @@ def display_header(self):
...
-            readline = None
-        if readline is not None:
-            print(f"{readline._READLINE_VERSION:#06x}")
-            is_editline = readline.__doc__ and "libedit" in readline.__doc__
-            print("editline" if is_editline else "GNU readline")
-        """
-        rl_output = subprocess.run([sys.executable, "-c", textwrap.dedent(rl_version_cmd)],
-                                   stdout=subprocess.PIPE, stderr=subprocess.PIPE)
-        if rl_output.returncode == 0:
+        rl_version_cmd = textwrap.dedent("""
```
Comment (nirs) [inline Lib/test/libregrtest/main.py]: The extra indent make this much more clear, thanks!

File: Lib/test/libregrtest/main.py [hunk 8]
```diff
@@ -433,26 +433,23 @@ def display_header(self):
...
-            except ValueError:
-                pass
+                import readline
+            except ImportError:
+                sys.exit(1)
             else:
-                print(f"== readline version: {rl_version}")
-                print(f"== readline implementation: {rl_implementation}")
+                def get_rl_info():
+                    is_editline = readline.__doc__ and "libedit" in readline.__doc__
```
Comment (nirs) [inline Lib/test/libregrtest/main.py]: This is now the third place using this horrible __doc__ check. I think we should eliminate this, now, or at least leave a todo and clean up later. Ideally we should have exactly one is_editline() function in cpython.

File: Lib/test/libregrtest/main.py [hunk 9]
```diff
@@ -433,26 +433,23 @@ def display_header(self):
...
+                    return f"{readline._READLINE_VERSION:#06x}", "editline" if is_editline else "GNU readline"
+                rl_version, rl_implementation = get_rl_info()
```
Comment (nirs) [inline Lib/test/libregrtest/main.py]: I think we should print the same values ptinted now in test_readline.setupModule

**Review by nirs — CHANGES_REQUESTED** [review]

File: Lib/test/libregrtest/main.py (hunk 5 above)
Comment (nirs) [reply Lib/test/libregrtest/main.py]: More like:

```python
def readline_info():
    script = """ 
import json
import readline
is_editline = readline.__doc__ and "libedit" in readline.__doc__
info = {"version": readline._READLINE_VERSION, 
        "implementation": "editline" if is_editline else "GNU readline"}
print(json.dumps(info))
"""
    c = subprocess.run([sys.executable, "-c", script],
                       stdout=subprocess.PIPE,
                       stderr=subprocess.PIPE)
    if c.returncode == 0:
        return json.loads(c.stdout.decode())
```

And use readline_info() in display_header(), or in any other context as needed.

**Comment by berkerpeksag** [comment]
Superseded by bpo-30871 and PR #3027.

**Comment by asottile** [comment]
@berkerpeksag  do you mean #3075?

**Comment by berkerpeksag** [comment]
@asottile yes that's what I meant! :) Good catch, thanks.

**Pull request closed without merging**